rpc/block_svc_routines.c

```diff
@@ -2469,6 +2475,332 @@ block_replace_cli_1_svc_st(blockReplaceCli *blk, struct svc_req *rqstp)
+      LOG("mgmt", GB_LOG_ERROR, "glfs_opendir(%s): on volume %s failed[%s]",
+          GB_METADIR, vols->data[i], strerror(errno));
+      ret = -1;
+      goto out;
+    }
+
+    while ((entry = glfs_readdir(tgmdfd))) {
+      if (strcmp(entry->d_name, ".") &&
+          strcmp(entry->d_name, "..") &&
+          strcmp(entry->d_name, "meta.lock")) {
+
+        if (GB_ALLOC(info) < 0) {
+          ret = -1;
+          goto out;
+        }
+        ret = blockGetMetaInfo(glfs, entry->d_name, info, NULL);
+        if (ret) {
+          goto out;
+        }
+
+        partOfBlock = false;
+        for (j = 0; j < info->nhosts; j++) {
+          if (blockhostIsValid(info->list[j]->status) && !strcmp(info->list[j]->addr, blk->addr)) {
+            partOfBlock = true;
+          }
+        }
+        if (!partOfBlock) {
+          continue;
+        }
+
+        /* storage_objects */
+        so_obj = getSoObj(entry->d_name, info, blk);
+        json_object_array_add(obj->so_arr, so_obj);
+
+        /* targets */
+        tg_obj = getTgObj(entry->d_name, info, blk);
+        json_object_array_add(obj->tg_arr, tg_obj);
+
+        blockFreeMetaInfo(info);
+      }
+    }
+
+    GB_METAUNLOCK(lkfd, vols->data[i], *errCode, *errMsg);
+    if (tgmdfd && glfs_closedir (tgmdfd) != 0) {
+      LOG("mgmt", GB_LOG_ERROR, "glfs_closedir(%s): on volume %s failed[%s]",
+          GB_METADIR, vols->data[i], strerror(errno));
+    }
+    if (lkfd && glfs_close(lkfd) != 0) {
+      LOG("mgmt", GB_LOG_ERROR, "glfs_close(%s): on volume %s failed[%s]",
+          GB_TXLOCKFILE, vols->data[i], strerror(errno));
+    }
+  }
+
+  ret = 0;
+  goto free;
+
+ out:
+  GB_METAUNLOCK(lkfd, vols->data[i], *errCode, *errMsg);
+  if (tgmdfd && glfs_closedir (tgmdfd) != 0) {
+    LOG("mgmt", GB_LOG_ERROR, "glfs_closedir(%s): on volume %s failed[%s]",
+        GB_METADIR, vols->data[i], strerror(errno));
+  }
+
+ optfail:
+  if (lkfd && glfs_close(lkfd) != 0) {
+    LOG("mgmt", GB_LOG_ERROR, "glfs_close(%s): on volume %s failed[%s]",
+        GB_TXLOCKFILE, vols->data[i], strerror(errno));
+  }
+
+ free:
+  strToCharArrayDefFree(vols);
+
+  return ret;
+}
+
+
+static int
+glusterBlockGenConfigSvc(blockGenConfigCli *blk,
+                         blockResponse *reply, char **errMsg, int *errCode)
+{
+  struct soTgObj *obj = NULL;
+  struct json_object *jobj = json_object_new_object();
+
+
+  if (GB_ALLOC(obj) < 0) {
+    goto out;
+  }
+  obj->so_arr = json_object_new_array();
+  obj->tg_arr = json_object_new_array();
+
+  reply->exit = getSoTgArraysForAllVolume(obj, blk, errMsg, errCode);
+  if(reply->exit) {
+    LOG("mgmt", GB_LOG_ERROR, "getSoTgArraysPerVolume(): on volume[s] %s failed",
+        blk->volume);
+    goto out;
+  }
+
+  json_object_object_add(jobj, "storage_objects", obj->so_arr);
+  json_object_object_add(jobj, "targets", obj->tg_arr);
+
+ out:
+  if(reply->exit) {
+    /* Do we need a json and non json output for failure ? */
```

Comment:
Shouldn't this be printed in json when --json is given?

Reply:
Right now we did not handled json for genconfig, was not sure if it is important.

Reply:
Keeping it here for archival purposes:

**Pranith:** How to find limit on the size of rpc response?


**Raghavendra Gowdappa:** There is no maximum limit on the rpc response. However, there is a limit on an individual rpc-fragment. RPC works around this limitation by making rpc message comprise of multiple fragments. But rpc traffic of glusterfs (from clients to bricks and vice-versa) have messages of only a single fragment. So, that way the maximum message size is the maximum rpc fragment which is 0x7fffffff (2047 MB).


**Pranith:** That is very big value for the stuff we are working on, so Prasanna, we are good.